src/Shelly.hs

```diff
@@ -1025,11 +1026,23 @@ instance Exception RunFailed
 
 show_command :: FilePath -> [Text] -> Text
 show_command exe args =
-    T.intercalate " " $ map quote (toTextIgnore exe : args)
-  where
-    quote t | T.any (== '\'') t = t
-    quote t | T.any isSpace t = surround '\'' t
-    quote t | otherwise = t
+  let escape char | char `Set.member` specialsInQuotes = T.pack ['\\', char]
+      escape char = T.singleton char
+      quote arg = surround '"' $ T.concatMap escape arg
+      isSafe c = all ($ c) [isPrint, not . isSpace, (`Set.notMember` specials)]
+      showArg "" = surround '"' ""
+      showArg arg | T.all isSafe arg = arg
+      showArg arg = quote arg
+  in T.intercalate " " $ map showArg (toTextIgnore exe : args)
+
+-- | Characters that need to be escaped or quoted to retain their literal value.
+specials :: Set.Set Char
+specials = Set.fromList "\\'\"`$&|;(){}<>"
+
+-- | When inside quotes, characters that need to be escaped to retain their
+-- literal value.
+specialsInQuotes :: Set.Set Char
+specialsInQuotes = Set.fromList $ ['\\', '"', '`', '$']
```

Comment:
Should this be conformed to the same style as `specials`?  (Optional, cosmetic change.)

Reply:
Made it a consistent style